File: pkg/dev_compiler/test/codegen_expected/lib/typed_data/float32x4_transpose_test.js
dart_library.library('lib/typed_data/float32x4_transpose_test', null, /* Imports */[
  'dart_sdk',
  'expect'
], function load__float32x4_transpose_test(exports, dart_sdk, expect) {
  'use strict';
  const core = dart_sdk.core;
  const typed_data = dart_sdk.typed_data;
  const dart = dart_sdk.dart;
  const dartx = dart_sdk.dartx;
  const expect$ = expect.expect;
  const float32x4_transpose_test = Object.create(null);
  let Float32x4ListTovoid = () => (Float32x4ListTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [typed_data.Float32x4List])))();
  let Float32x4ListAndFloat32x4ListTovoid = () => (Float32x4ListAndFloat32x4ListTovoid = dart.constFn(dart.definiteFunctionType(dart.void, [typed_data.Float32x4List, typed_data.Float32x4List])))();
  let VoidTodynamic = () => (VoidTodynamic = dart.constFn(dart.definiteFunctionType(dart.dynamic, [])))();
  float32x4_transpose_test.transpose = function(m) {
    expect$.Expect.equals(4, m.length);
    let m0 = m.get(0);
    let m1 = m.get(1);
    let m2 = m.get(2);
    let m3 = m.get(3);
    let t0 = m0.shuffleMix(m1, typed_data.Float32x4.XYXY);
    let t1 = m2.shuffleMix(m3, typed_data.Float32x4.XYXY);
    m.set(0, t0.shuffleMix(t1, typed_data.Float32x4.XZXZ));
    m.set(1, t0.shuffleMix(t1, typed_data.Float32x4.YWYW));
    let t2 = m0.shuffleMix(m1, typed_data.Float32x4.ZWZW);
    let t3 = m2.shuffleMix(m3, typed_data.Float32x4.ZWZW);
    m.set(2, t2.shuffleMix(t3, typed_data.Float32x4.XZXZ));
    m.set(3, t2.shuffleMix(t3, typed_data.Float32x4.YWYW));
  };
  dart.fn(float32x4_transpose_test.transpose, Float32x4ListTovoid());
  float32x4_transpose_test.testTranspose = function(m, r) {
    float32x4_transpose_test.transpose(m);
    for (let i = 0; i < 4; i++) {
      let a = m.get(i);
      let b = r.get(i);
      expect$.Expect.equals(b.x, a.x);
      expect$.Expect.equals(b.y, a.y);
      expect$.Expect.equals(b.z, a.z);
      expect$.Expect.equals(b.w, a.w);
    }
  };
  dart.fn(float32x4_transpose_test.testTranspose, Float32x4ListAndFloat32x4ListTovoid());
  float32x4_transpose_test.main = function() {
    let A = typed_data.Float32x4List.new(4);
    A.set(0, typed_data.Float32x4.new(1.0, 2.0, 3.0, 4.0));
    A.set(1, typed_data.Float32x4.new(5.0, 6.0, 7.0, 8.0));
    A.set(2, typed_data.Float32x4.new(9.0, 10.0, 11.0, 12.0));
    A.set(3, typed_data.Float32x4.new(13.0, 14.0, 15.0, 16.0));
    let B = typed_data.Float32x4List.new(4);
    B.set(0, typed_data.Float32x4.new(1.0, 5.0, 9.0, 13.0));
    B.set(1, typed_data.Float32x4.new(2.0, 6.0, 10.0, 14.0));
    B.set(2, typed_data.Float32x4.new(3.0, 7.0, 11.0, 15.0));
    B.set(3, typed_data.Float32x4.new(4.0, 8.0, 12.0, 16.0));
    let I = typed_data.Float32x4List.new(4);
    I.set(0, typed_data.Float32x4.new(1.0, 0.0, 0.0, 0.0));
    I.set(1, typed_data.Float32x4.new(0.0, 1.0, 0.0, 0.0));
    I.set(2, typed_data.Float32x4.new(0.0, 0.0, 1.0, 0.0));
    I.set(3, typed_data.Float32x4.new(0.0, 0.0, 0.0, 1.0));
    for (let i = 0; i < 20; i++) {
      let m = typed_data.Float32x4List.fromList(I);
      float32x4_transpose_test.testTranspose(m, I);
      m = typed_data.Float32x4List.fromList(A);
      float32x4_transpose_test.testTranspose(m, B);
    }
  };
  dart.fn(float32x4_transpose_test.main, VoidTodynamic());
  // Exports:
  exports.float32x4_transpose_test = float32x4_transpose_test;
});

File: tests/lib_2/typed_data/float32x4_transpose_test.dart
// Copyright (c) 2013, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.
// VMOptions=--optimization-counter-threshold=10 --no-background-compilation

// Library tag to be able to run in html test framework.
library float32x4_transpose_test;

import 'dart:typed_data';
import 'package:expect/expect.dart';

void transpose(Float32x4List m) {
  Expect.equals(4, m.length);
  var m0 = m[0];
  var m1 = m[1];
  var m2 = m[2];
  var m3 = m[3];

  var t0 = m0.shuffleMix(m1, Float32x4.xyxy);
  var t1 = m2.shuffleMix(m3, Float32x4.xyxy);
  m[0] = t0.shuffleMix(t1, Float32x4.xzxz);
  m[1] = t0.shuffleMix(t1, Float32x4.ywyw);

  var t2 = m0.shuffleMix(m1, Float32x4.zwzw);
  var t3 = m2.shuffleMix(m3, Float32x4.zwzw);
  m[2] = t2.shuffleMix(t3, Float32x4.xzxz);
  m[3] = t2.shuffleMix(t3, Float32x4.ywyw);
}

void testTranspose(Float32x4List m, Float32x4List r) {
  transpose(m); // In place transpose.
  for (int i = 0; i < 4; i++) {
    var a = m[i];
    var b = r[i];
    Expect.equals(b.x, a.x);
    Expect.equals(b.y, a.y);
    Expect.equals(b.z, a.z);
    Expect.equals(b.w, a.w);
  }
}

main() {
  var A = new Float32x4List(4);
  A[0] = new Float32x4(1.0, 2.0, 3.0, 4.0);
  A[1] = new Float32x4(5.0, 6.0, 7.0, 8.0);
  A[2] = new Float32x4(9.0, 10.0, 11.0, 12.0);
  A[3] = new Float32x4(13.0, 14.0, 15.0, 16.0);
  var B = new Float32x4List(4);
  B[0] = new Float32x4(1.0, 5.0, 9.0, 13.0);
  B[1] = new Float32x4(2.0, 6.0, 10.0, 14.0);
  B[2] = new Float32x4(3.0, 7.0, 11.0, 15.0);
  B[3] = new Float32x4(4.0, 8.0, 12.0, 16.0);
  var I = new Float32x4List(4);
  I[0] = new Float32x4(1.0, 0.0, 0.0, 0.0);
  I[1] = new Float32x4(0.0, 1.0, 0.0, 0.0);
  I[2] = new Float32x4(0.0, 0.0, 1.0, 0.0);
  I[3] = new Float32x4(0.0, 0.0, 0.0, 1.0);
  for (int i = 0; i < 20; i++) {
    var m = new Float32x4List.fromList(I);
    testTranspose(m, I);
    m = new Float32x4List.fromList(A);
    testTranspose(m, B);
  }
}
